Expose Keycloak config errors in the Keycloak CR status field

### Description

At startup, Keycloak performs several config validations ensuring all required options are specified and that the options values have correct format. In case the validations fail, an exception is thrown. Exceptions are also thrown in other misconfiguration cases such as wrong DB credentials.

The Operator should be able to expose this kind of errors and expose them in the CR's status.

We should avoid implementing server config validations at the Operator level as the Operator does not know what config options has been already preset in a Keycloak custom image. Server config validations need to be done at the server level.

### Discussion

_No response_

### Motivation

_No response_

### Details

Implementation requires discussion.

We might need to try and interpret the server log and look for known exception types. Other way might be to send some message with failure details to the Operator (a new log handler?).

We might also need to improve the exceptions format (and possibly types) for the startup failures due to misconfiguration.

Without the verbose option the errors seem to appear on a single line:

```
ERROR: Strict hostname resolution configured but no hostname setting provided
```

With the verbose option the message is embedded in the stacktrace:

```
ERROR: Unexpected error when starting the server in (production) mode
Error details:
java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
...
Caused by: java.lang.RuntimeException: Strict hostname resolution configured but no hostname setting provided
        at org.keycloak.quarkus.runtime.hostname.DefaultHostnameProvider.init(DefaultHostnameProvider.java:238)
        at org.keycloak.quarkus.runtime.integration.QuarkusKeycloakSessionFactory.<init>(QuarkusKeycloakSessionFactory.java:83)
```

> We might need to try and interpret the server log and look for known exception types.

We could scrape the pod logs in KeycloakDeployment.checkForPodErrors.

Would that require users of pod template overrides specifying the start command also supply the verbose option to make this functionality work, or would try to account for both verbose and non-verbose logs?

> Other way might be to send some message with failure details to the Operator (a new log handler?).

Seems like we could use kubernetes events for this.  The only drawback would be the short retention time, but once added to the keycloak status they could be retained if the status aggregation logic knows to not clear them.

> We might also need to improve the exceptions format (and possibly types) for the startup failures due to misconfiguration.

That seems good too.  It seems a little problematic to parse a semi-structured format and scraping for exception messages.

--- a/operator/src/main/java/org/keycloak/operator/controllers/KeycloakController.java
+++ b/operator/src/main/java/org/keycloak/operator/controllers/KeycloakController.java
@@ -22,6 +22,7 @@ import io.fabric8.kubernetes.api.model.ContainerStatus;
 import io.fabric8.kubernetes.api.model.PodSpec;
 import io.fabric8.kubernetes.api.model.PodStatus;
 import io.fabric8.kubernetes.api.model.apps.StatefulSet;
+import io.fabric8.kubernetes.client.KubernetesClientException;
 import io.fabric8.kubernetes.client.readiness.Readiness;
 import io.fabric8.kubernetes.client.utils.KubernetesResourceUtil;
 import io.fabric8.kubernetes.client.utils.Serialization;
@@ -284,6 +285,14 @@ public class KeycloakController implements Reconciler<Keycloak> {
                                 if (Optional.ofNullable(cs.getState()).map(ContainerState::getWaiting)
                                         .map(ContainerStateWaiting::getReason).map(String::toLowerCase)
                                         .filter(s -> s.contains("err") || s.equals("crashloopbackoff")).isPresent()) {
+                                    // since we've failed, try to get the previous first, then the current
+                                    String log = null;
+                                    try {
+                                        log = context.getClient().raw(String.format("/api/v1/namespaces/%s/pods/%s/log?previous=true&tailLines=200", p.getMetadata().getNamespace(), p.getMetadata().getName()));
+                                    } catch (KubernetesClientException e) {
+                                        // just ignore
+                                    }
+
                                     Log.infof("Found unhealthy container on pod %s/%s: %s",
                                             p.getMetadata().getNamespace(), p.getMetadata().getName(),
                                             Serialization.asYaml(cs));
@@ -291,6 +300,14 @@ public class KeycloakController implements Reconciler<Keycloak> {
                                             String.format("Waiting for %s/%s due to %s: %s", p.getMetadata().getNamespace(),
                                                     p.getMetadata().getName(), cs.getState().getWaiting().getReason(),
                                                     cs.getState().getWaiting().getMessage()));
+                                    if (log != null) {
+                                        if (log.length() > 2000) {
+                                            log = "... " + log.substring(log.length() - 2000, log.length());
+                                        }
+                                        status.addErrorMessage(
+                                                String.format("Log for %s/%s: %s", p.getMetadata().getNamespace(),
+                                                        p.getMetadata().getName(), log));
+                                    }
                                 }
                             });
                 });

--- a/operator/src/main/kubernetes/kubernetes.yml
+++ b/operator/src/main/kubernetes/kubernetes.yml
@@ -43,6 +43,12 @@ rules:
     verbs:
       - list
   - apiGroups:
+      - ""
+    resources:
+      - pods/log
+    verbs:
+      - get
+  - apiGroups:
       - batch
     resources:
       - jobs

--- a/operator/src/test/java/org/keycloak/operator/testsuite/integration/KeycloakDeploymentTest.java
+++ b/operator/src/test/java/org/keycloak/operator/testsuite/integration/KeycloakDeploymentTest.java
@@ -32,6 +32,7 @@ import io.fabric8.kubernetes.client.dsl.Resource;
 import io.quarkus.logging.Log;
 import io.quarkus.test.junit.QuarkusTest;
 
+import org.assertj.core.api.Condition;
 import org.awaitility.Awaitility;
 import org.junit.jupiter.api.Assumptions;
 import org.junit.jupiter.api.Tag;
@@ -46,6 +47,7 @@ import org.keycloak.operator.crds.v2alpha1.deployment.Keycloak;
 import org.keycloak.operator.crds.v2alpha1.deployment.KeycloakStatusCondition;
 import org.keycloak.operator.crds.v2alpha1.deployment.ValueOrSecret;
 import org.keycloak.operator.crds.v2alpha1.deployment.spec.BootstrapAdminSpec;
+import org.keycloak.operator.crds.v2alpha1.deployment.spec.FeatureSpecBuilder;
 import org.keycloak.operator.crds.v2alpha1.deployment.spec.HostnameSpecBuilder;
 import org.keycloak.operator.testsuite.apiserver.DisabledIfApiServerTest;
 import org.keycloak.operator.testsuite.unit.WatchedResourcesTest;
@@ -542,6 +544,25 @@ public class KeycloakDeploymentTest extends BaseOperatorTest {
             CRAssert.assertKeycloakStatusCondition(current, KeycloakStatusCondition.HAS_ERRORS, true,
                     String.format("Waiting for %s/%s-0 due to ErrImage", k8sclient.getNamespace(),
                             kc.getMetadata().getName()));
+        });
+    }
+
+    @Test
+    public void testConfigErrorLog() {
+        var kc = getTestKeycloakDeployment(true);
+        kc.getSpec().setFeatureSpec(new FeatureSpecBuilder().addToEnabledFeatures("feature doesn't exist").build());
+
+        deployKeycloak(k8sclient, kc, false);
+
+        var crSelector = k8sclient.resource(kc);
+
+        Awaitility.await().atMost(3, MINUTES).pollDelay(1, SECONDS).ignoreExceptions().untilAsserted(() -> {
+            Keycloak current = crSelector.get();
+            CRAssert.assertKeycloakStatusCondition(current, KeycloakStatusCondition.READY, false);
+            CRAssert.assertKeycloakStatusCondition(current, KeycloakStatusCondition.HAS_ERRORS, true, null).has(new Condition<>(
+                    c -> c.getMessage().contains(String.format("Waiting for %s/%s-0 due to CrashLoopBackOff", k8sclient.getNamespace(), kc.getMetadata().getName()))
+                     && c.getMessage().contains("feature doesn't exist"), "message"
+                    ));
         });
     }
 

--- a/operator/src/test/java/org/keycloak/operator/testsuite/utils/CRAssert.java
+++ b/operator/src/test/java/org/keycloak/operator/testsuite/utils/CRAssert.java
@@ -75,10 +75,10 @@ public final class CRAssert {
     public static void assertKeycloakStatusCondition(Keycloak kc, String condition, Boolean status) {
         assertKeycloakStatusCondition(kc, condition, status, null);
     }
-    public static void assertKeycloakStatusCondition(Keycloak kc, String condition, Boolean status, String containedMessage) {
+    public static ObjectAssert<KeycloakStatusCondition> assertKeycloakStatusCondition(Keycloak kc, String condition, Boolean status, String containedMessage) {
         Log.debugf("Asserting CR: %s, condition: %s, status: %s, message: %s", kc.getMetadata().getName(), condition, status, containedMessage);
         try {
-            assertKeycloakStatusCondition(kc.getStatus(), condition, status, containedMessage, null);
+            return assertKeycloakStatusCondition(kc.getStatus(), condition, status, containedMessage, null);
         } catch (Exception e) {
             Log.infof("Asserting CR: %s with status:\n%s", kc.getMetadata().getName(), Serialization.asYaml(kc.getStatus()));
             throw e;

